File: Week 3/Python Codes/RotateLinkedList.py

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# This function should rotate list counter-
# clockwise by k and return new head (if changed) 
def rotateList(head, k):
    kth_node = head
    last_node = head
    count = 1
    while kth_node.next is not None and count < k:
        kth_node = kth_node.next
        count += 1

    while last_node.next is not None:
        last_node = last_node.next
    
    last_node.next = head
    head = kth_node.next
    kth_node.next = None
    return head

File: Week 3/C++ Codes/RotateLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    struct Node *next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};

// This function should rotate list counter-clockwise
// by k and return new head (if changed)
Node* rotate(Node* head, int k)
{
    int counter = 1;
    struct Node* last = head;
    struct Node* kth = head;
    while (counter < k) {
        kth = kth->next;        
        counter += 1;
    }
    while (last->next != NULL) {
        last = last->next;
    }
    last->next = head;
    head = kth->next;
    kth->next = NULL;
    return head;
}
